orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -326,13 +327,20 @@ private static TypeDescription buildOrcProjection(
             orcType = originalType.clone();
           }
         } else {
+          Types.NestedField field = root.findField(fieldId);
           if (isRequired) {
-            throw new IllegalArgumentException(
+            Preconditions.checkArgument(
+                field.initialDefault() != null,
```

Comment:
Minor: Do we even need to check that the field.initialDefault() != null? If we're in this block (the field does not exist in the file) and the field is required, couldn't we just throw the IllegalArgumentException?

```
Preconditions.checkArgument(!isRequired, ""Missing required field: %s (%s)")
```

Let me know if I'm missing anything.

Agree with including the full column name though, that's a good improvement.

Reply:
>This will also not fail missing required fields that have defaults. Instead, the UnsupportedOperationException for default support is thrown.

Ah I see the intent behind this, if we have missing required field that is set with defaults we do want to fall through to the new unsupported operation failure. I think it makes sense to indicate a different failure, and in that case the first check should check `field.initialDefaul() != null`